server/src/main/java/org/elasticsearch/search/internal/SearchContext.java

```diff
@@ -374,64 +759,672 @@ public void clearReleasables(Lifetime lifetime) {
+                aggregations = new SearchContextAggregations(factories, multiBucketConsumer);
+            } catch (IOException e) {
+                throw new AggregationInitializationException("Failed to create aggregators", e);
+            }
+            return this;
+        }
+
         /**
-         * This life time is for objects that need to live until the search context they are attached to is destroyed.
+         * Builds the script fields context from the provided builder.
          */
-        CONTEXT
+        public Builder buildScriptFields(ScriptService scriptService,
+                                         Collection<SearchSourceBuilder.ScriptField> fields) {
+            int maxAllowedScriptFields = queryShardContext.getIndexSettings().getMaxScriptFields();
+            if (fields.size() > maxAllowedScriptFields) {
+                throw new IllegalArgumentException(
+                    "Trying to retrieve too many script_fields. Must be less than or equal to: [" + maxAllowedScriptFields
+                        + "] but was [" + fields.size() + "]. This limit can be set by changing the ["
+                        + IndexSettings.MAX_SCRIPT_FIELDS_SETTING.getKey() + "] index level setting.");
+            }
+            List<ScriptFieldsContext.ScriptField> result = new ArrayList<>();
+            for (SearchSourceBuilder.ScriptField field : fields) {
+                FieldScript.Factory factory = scriptService.compile(field.script(), FieldScript.CONTEXT);
+                FieldScript.LeafFactory searchScript = factory.newFactory(field.script().getParams(), queryShardContext.lookup());
+                result.add(new ScriptFieldsContext.ScriptField(field.fieldName(), searchScript, field.ignoreFailure()));
+            }
+            this.scriptFields = new ScriptFieldsContext(result);
+            return this;
+        }
+
+        /**
+         * Builds the doc values context from the provided list of fields.
+         */
+        public Builder buildDocValues(List<DocValueFieldsContext.FieldAndFormat> fields) {
+            List<DocValueFieldsContext.FieldAndFormat> docValueFields = new ArrayList<>();
+            for (DocValueFieldsContext.FieldAndFormat format : fields) {
+                Collection<String> fieldNames = queryShardContext.getMapperService().simpleMatchToFullName(format.field);
+                for (String fieldName : fieldNames) {
+                    docValueFields.add(new DocValueFieldsContext.FieldAndFormat(fieldName, format.format));
+                }
+            }
+            this.docValueFieldsContext = new DocValueFieldsContext(docValueFields);
+            return this;
+        }
+
+        /**
+         * Creates the final {@link SearchContext} and set the provided
+         * {@link Runnable} to be executed when the context is closed.
+         */
+        public SearchContext build(Runnable onClose) throws SearchException {
```

Comment:
Is there a reason why we take `onClose` as a parameter here, and don't set it on the Builder directly? Almost every invocation seems to be `build(() -> {})` which suggests that it's not necessary.